doc/document.go

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package doc
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"sort"
+	"unicode/utf8"
+)
+
+var (
+	errMultipleIDs   = errors.New("document cannot contain multiple IDs")
+	errZeroLengthID  = errors.New("document ID cannot be of length zero")
+	errEmptyDocument = errors.New("document cannot be empty")
+)
+
+// IDReservedFieldName is the field name reserved for IDs.
+var IDReservedFieldName = []byte("_id")
```

Comment:
wanted to you intended this is OK to be changed by a downstream library user. that's good yea?

Reply:
No, I don't think this should be ever be changed, but it can be if the caller so chooses. It's buyer-beware though, we don't offer any guarantees if they do.

Reply:
`_m3ninx_id`